clients/src/main/java/org/apache/kafka/clients/admin/ListConsumerGroupsResult.java

```diff
@@ -29,16 +34,75 @@
  */
 @InterfaceStability.Evolving
 public class ListConsumerGroupsResult {
-    private final KafkaFuture<Collection<ConsumerGroupListing>> future;
+    private final Map<Node, KafkaFutureImpl<Collection<ConsumerGroupListing>>> futuresMap;
+    private final KafkaFuture<Collection<ConsumerGroupListing>> flattenFuture;
+    private final KafkaFuture<Void> listFuture;
 
-    ListConsumerGroupsResult(KafkaFuture<Collection<ConsumerGroupListing>> future) {
-        this.future = future;
+    ListConsumerGroupsResult(final KafkaFuture<Void> listFuture,
+                             final KafkaFuture<Collection<ConsumerGroupListing>> flattenFuture,
+                             final Map<Node, KafkaFutureImpl<Collection<ConsumerGroupListing>>> futuresMap) {
+        this.flattenFuture = flattenFuture;
+        this.listFuture = listFuture;
+        this.futuresMap = futuresMap;
+    }
+
+    private class FutureConsumerGroupListingIterator extends AbstractIterator<KafkaFuture<ConsumerGroupListing>> {
+        private Iterator<KafkaFutureImpl<Collection<ConsumerGroupListing>>> futuresIter;
+        private Iterator<ConsumerGroupListing> innerIter;
+
+        @Override
+        public boolean hasNext() {
```

Comment:
It's a little weird that we're overriding this. I would have expected we would do all this in `makeNext`

Reply:
ack.